### PR TITLE
Add support for 74x595 shift registers

### PR DESCRIPTION
**Pull request by jziolkowski**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [x] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.1.1
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

File: tasmota/i18n.h [hunk 1]
```diff
@@ -716,6 +716,11 @@
 #define D_CMND_BR_RUN ""
 #define D_BR_NOT_STARTED  "Berry not started"
 
+// Commands xdrv_60_shift595.ino - 74x595 family shift register driver
```
Comment (barbudor) [inline tasmota/i18n.h:719]: No need to #ifdef, just let it defined any time as it takes no memory

File: tasmota/my_user_config.h [hunk 2]
```diff
@@ -350,6 +350,8 @@
 #define ZIGBEE_DISTINCT_TOPICS false             // [SetOption89] Enable unique device topic based on Zigbee device ShortAddr
 #define ZIGBEE_RMV_ZBRECEIVED  false             // [SetOption100] Remove ZbReceived form JSON message
 #define ZIGBEE_INDEX_EP        false             // [SetOption101] Add the source endpoint as suffix to attributes, ex `Power3` instead of `Power` if sent from endpoint 3
+#define SHIFT595_INVERT_OUTPUTS false            // [SetOption133] Don't invert outputs of 74x595 shift register
+#define SHIFT595_DEVICE_COUNT  1                 // [Shift595DeviceCount] Set the number of connected 74x595 shift registers
```
Comment (barbudor) [inline tasmota/my_user_config.h:354]: I think those should be located right below USE_SHIFT595 and indented by 2 spaces to let people understand they are option of the USE_595 driver

File: tasmota/tasmota_template.h [hunk 3]
```diff
@@ -372,7 +373,11 @@ const char kSensorNames[] PROGMEM =
   D_SENSOR_BL0942_RX "|"
   D_SENSOR_HM330X_SET "|"
   D_SENSOR_HEARTBEAT "|" D_SENSOR_HEARTBEAT "_i|"
-  ;
+
+#ifdef USE_SHIFT595
+  D_GPIO_SHIFT595_SRCLK "|" D_GPIO_SHIFT595_RCLK "|" D_GPIO_SHIFT595_OE "|" D_GPIO_SHIFT595_SER "|"
+#endif
+;
```
Comment (barbudor) [inline tasmota/tasmota_template.h:380]: No #ifdef for those, htey must always be there, like GPIO_SHIFT595_SRCLK, GPIO_SHIFT595_RCLK, GPIO_SHIFT595_OE, GPIO_SHIFT595_SER

File: tasmota/xdrv_60_shift595.ino [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+struct Shift595 {
+  int8_t pinSRCLK;
+  int8_t pinRCLK;
+  int8_t pinSER;
+  int8_t pinOE;
+  int8_t outputs;
+  int8_t first = TasmotaGlobal.devices_present;
+  bool connected = false;
+} Shift595;
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:34]: To save RAM when the driver is included in a build but used change for 
```
} *Shift595 = nullptr;
```
Change all your `Shift595.`to `Shift595->`
Allocate `Shift595 = calloc(1 sizeof(struct Shift595));` when initializing hte driver (if GPIOS are defined)
Test `if (Shift595)` before calling any function that need to use `Shift595->`

File: tasmota/xdrv_60_shift595.ino [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+const char kShift595Commands[] PROGMEM = "|" D_CMND_SHIFT595_DEVICE_COUNT ;
+void (* const Shift595Command[])(void) PROGMEM = { &CmndShift595Devices };
+
+struct Shift595 {
+  int8_t pinSRCLK;
+  int8_t pinRCLK;
+  int8_t pinSER;
+  int8_t pinOE;
+  int8_t outputs;
+  int8_t first = TasmotaGlobal.devices_present;
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:32]: That won't work.
You must set `Shift595.first = TasmotaGlobal.devices_present` in your driver init function

File: tasmota/xdrv_60_shift595.ino [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  bool connected = false;
+} Shift595;
+
+void Shift595Init(void)
+{
+  if (PinUsed(GPIO_SHIFT595_SRCLK) && PinUsed(GPIO_SHIFT595_RCLK) && PinUsed(GPIO_SHIFT595_SER)) {
+    Shift595.pinSRCLK = Pin(GPIO_SHIFT595_SRCLK);
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:39]: This is where you can allocate memory for Shift595 just after the `if`

File: tasmota/xdrv_60_shift595.ino [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    pinMode(Shift595.pinSER, OUTPUT);
+    digitalWrite(Shift595.pinSER, 0);
+    
+    if (PinUsed(GPIO_SHIFT595_OE)) {
+      Shift595.pinOE = Pin(GPIO_SHIFT595_OE);
+      pinMode(Shift595.pinOE, OUTPUT);
+      digitalWrite(Shift595.pinOE, 1);     
+    }
+    Shift595.outputs = Settings->shift595_device_count * 8;
+    TasmotaGlobal.devices_present += Shift595.outputs;
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:57]: This is where you should save `devices_present` into `SHift595->first`, right before incrementing `devices_present`

File: tasmota/xdrv_60_shift595.ino [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+
+bool Xdrv60(uint8_t function)
+{
+  bool result = false;
+  switch (function) {
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:100]: See https://github.com/arendst/Tasmota/blob/4d05fb153c1e67cbdc83dac18c002f4b0c57e7f6/tasmota/xsns_93_hm330x.ino#L325-L353
How to manage this when using a pointer for the structure so driver's function are not called if Init hasn't allocated the structure

File: tasmota/xdrv_60_shift595.ino [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#ifdef USE_SHIFT595
+
+#define XDRV_60           60
+
+const char kShift595Commands[] PROGMEM = "|" D_CMND_SHIFT595_DEVICE_COUNT ;
+void (* const Shift595Command[])(void) PROGMEM = { &CmndShift595Devices };
```
Comment (barbudor) [inline tasmota/xdrv_60_shift595.ino:24]: If you use `Shift595` as a prefix, that will open the door to more commands :
```
const char kShift595Commands[] PROGMEM = D_PRFX_SHIFT595 "|" D_CMND_SHIFT595_DEVICE_COUNT ;
```
With
```
#define D_PRFX_SHIFT595 "Shift595"
#define D_CMND_SHIFT595_DEVICE_COUNT "Devices"
```

Comment (jziolkowski) [reply tasmota/xdrv_60_shift595.ino:24]: yes, I did consider that but with one command currently it's pointless. Can be always added.

**Review by barbudor — APPROVED** [review]
👍

**Pull request merged**